allure-support/src/main/kotlin/com/kaspersky/components/alluresupport/interceptors/testrun/MoveReportsInterceptor.kt

```diff
@@ -0,0 +1,124 @@
+    private val instrumentation: Instrumentation,
+    private val dirsProvider: DirsProvider,
+    private val rootDirsProvider: ResourcesRootDirsProvider,
+    private val stateHolder: TestRunStateHolder,
+    private val device: UiDevice
+) : TestRunWatcherInterceptor {
+    override fun onTestFinished(testInfo: TestInfo, success: Boolean) {
+        val allureTargetDir = moveAllureReportToSdCard()
+        removeStubs(allureTargetDir)
+        moveAttachedVideosToProperDirectories(allureTargetDir)
+        cleanUp()
+    }
+
+    /**
+     * Deletes allure report under /data/data/your.package.name/files and empty directories after moving real videos from them
+     */
+    private fun cleanUp() {
+        getOriginalAllureDir().deleteRecursively()
+        val videosDir = dirsProvider.provideNew(rootDirsProvider.videoRootDir)
+        device.executeShellCommand("rm -rf ${videosDir.absolutePath}")
+
+        dirsProvider.provideNew(rootDirsProvider.screenshotsRootDir).deleteRecursively()
+        dirsProvider.provideNew(rootDirsProvider.logcatRootDir).deleteRecursively()
+        dirsProvider.provideNew(rootDirsProvider.viewHierarchy).deleteRecursively()
+    }
+
+    /**
+     * @return allure results dir under /data/data/your.package.name/files
+     */
+    private fun getOriginalAllureDir(): File = instrumentation.targetContext.filesDir.resolve("allure-results")
+
+    /**
+     * Moves allure report from /data/data/your.package.name/files/allure-report to external storage e.g.
+     * /sdcard/Documents/.../allure-results
+     * @return allure target directory on sdcard
+     */
+    private fun moveAllureReportToSdCard(): File {
+        val allureResultsFile = getOriginalAllureDir()
+        val allureTargetDir = dirsProvider.provideNew(rootDirsProvider.allureRootDir)
+        if (!allureResultsFile.exists()) {
+            throw IllegalArgumentException("Unable to move allure results from $allureResultsFile. File not found")
+        }
+        allureResultsFile.copyRecursively(allureTargetDir)
+
+        return allureTargetDir
+    }
+
+    /**
+     * During allure test stubs are attached to report because real videos can't be recorded directly to /data/data.
+     * After kaspresso moves report to sdcard it has to replace stubs with real videos. Videos are moved using adb shell
+     * 👀 https://issuetracker.google.com/issues/258277873
+     * @param allureTargetDir allure directory under /sdcard
+     */
+    private fun removeStubs(allureTargetDir: File) {
+        allureTargetDir.listFiles()?.forEach {
+            if (it.extension.contains(MP4_EXTENSION, ignoreCase = true)) {
+                it.delete()
+            }
+        }
+    }
+
+    /**
+     * Move (and rename) real videos from their original directories to report dir
+     * @param allureTargetDir allure directory under /sdcard
+     */
+    private fun moveAttachedVideosToProperDirectories(allureTargetDir: File) {
+        stateHolder.attachedVideos.forEach {
```

Comment:
To be honest, I don't really like this solution.
It seems unobvious to me. But what if there was no call to stateHolder before? Another interceptor that works with stateHolder may not be connected.
Maybe it makes sense to do it with one interceptor? To ensure both are connected at the same time

Reply:
Good point. I'll make this interceptor remember uuid when test is started, so there would be no need in additional one. Move interceptor need to be a separate one though because it let us move reports even if video was not recorded. In this case we just move report without any videos. If we merge video and move interceptor we won't be able to move reports from /data/data